src/cpp/lpnamer/main/RunProblemGeneration.cpp

```diff
@@ -92,15 +100,139 @@ void RunProblemGeneration(
   Couplings couplings;
   std::string solver_name = "CBC";
   std::vector<ActiveLink> links = linkBuilder.getLinks();
+
+  auto const mps_file_name = xpansion_output_dir / common_lpnamer::MPS_TXT;
+
   LinkProblemsGenerator linkProblemsGenerator(links, solver_name, logger,
                                               log_file_path, zip_mps);
-  linkProblemsGenerator.treatloop(xpansion_output_dir, antares_archive_path,
-                                  couplings);
+  auto files_mapper = FilesMapper(antares_archive_path);
+  auto mpsList = files_mapper.MpsAndVariablesFilesVect();
+
+  auto lpDir_ = xpansion_output_dir / "lp";
+
+  bool use_zip_implementation = true;
+  bool use_file_implementation = false;
+  if (use_zip_implementation) {
+    std::shared_ptr<ArchiveReader> reader =
+        InstantiateZipReader(antares_archive_path);
+    std::shared_ptr<ArchiveWriter> writer;
+    const std::filesystem::path tmpArchivePath = OpenTmpArchive(lpDir_, writer);
+
+    /* Zip problem writer */
+    auto problem_writer =
+        std::make_shared<ArchiveProblemWriter>(lpDir_, writer);
+
+    /* Main stuff */
+    std::vector<std::string> problem_names;
+    std::transform(mpsList.begin(), mpsList.end(),
+                   std::back_inserter(problem_names),
+                   [](ProblemData const& data) { return data._problem_mps; });
+    auto adapter = std::make_shared<ZipProblemsProviderAdapter>(lpDir_, reader,
+                                                                problem_names);
+    std::vector<std::shared_ptr<Problem>> xpansion_problems =
+        adapter->provideProblems(solver_name);
+
+    std::vector<std::pair<std::shared_ptr<Problem>, ProblemData>>
+        problems_and_data;
+    for (int i = 0; i < xpansion_problems.size(); ++i) {
+      xpansion_problems.at(i)->_name = mpsList.at(i)._problem_mps;
+      problems_and_data.emplace_back(xpansion_problems.at(i), mpsList.at(i));
+    }
+    std::for_each(std::execution::par, problems_and_data.begin(),
+                  problems_and_data.end(), [&](const auto& problem_and_data) {
+                    const auto& [problem, data] = problem_and_data;
+
+                    auto problem_variables_from_zip_adapter =
+                        std::make_shared<ProblemVariablesZipAdapter>(
+                            reader, data, links, logger);
+                    linkProblemsGenerator.treat(
+                        data._problem_mps, couplings, problem_writer, problem,
+                        problem_variables_from_zip_adapter);
+                  });
```

Comment:
it is possible to merge the "for" and "for_each" loops

Reply:
Not exactly.
I need the ascociation of "problem" and "data" both contained in two different vectors: xpansion_problems and mpsList.
You are right that I could iterate through both vectors at the same time using one for loop with the same index.
However my goal with the for_each loop is to parallelize the operations and I can't itarate both vector at once with a for_each loop nor use a index based loop.